include/render.hpp

```diff
@@ -143,40 +148,39 @@ void render(sycl::queue& queue, std::array<color, width * height>& fb,
   const auto ideal_pg_height = height / parallel_exec;
   const auto pg_height = (ideal_pg_height == 0) ? 1 : ideal_pg_height;
   const auto pg_width =
-	  (ideal_pg_height == 0)
-		  ? find_immediate_least_divider<width>(total_work / parallel_exec)
-		  : width;
+      (ideal_pg_height == 0)
+          ? find_immediate_least_divider<width>(total_work / parallel_exec)
+          : width;
   const auto nb_pg_line = width / pg_width;
   const auto required_pg_line =
-	  height / ideal_pg_height + ((height % ideal_pg_height) ? 1 : 0);
+      height / ideal_pg_height + ((height % ideal_pg_height) ? 1 : 0);
   const auto required_pg = required_pg_line * nb_pg_line;
 
   // Submit command group on device
   queue.submit([=, &hittables_buf, &frame_buf, &cam](sycl::handler& cgh) {
-	auto fb_acc = frame_buf.get_access<sycl::access::mode::discard_write>(cgh);
-	auto hittables_acc =
-		hittables_buf.get_access<sycl::access::mode::read>(cgh);
-	hittable_t const* hittable_ptr = hittables_acc.get_pointer();
-	color* fb_ptr = fb_acc.get_pointer();
-	const auto global = sycl::range<1>(required_pg);
-	const auto local = sycl::range<1>(wg_size);
-	cgh.parallel_for_work_group(global, local, [=](sycl::group<1> g) {
-	  g.parallel_for_work_item([&](sycl::h_item<1> item) {
-		auto gid = item.get_global_id()[0];
-		const auto grid_x = gid % nb_pg_line;
-		const auto grid_y = gid / nb_pg_line;
-		const auto start_x = grid_x * pg_width;
-		const auto start_y = grid_y * pg_height;
-		const auto max_x = start_x + pg_width;
-		const auto th_max_y = start_y + pg_height;
-		const auto max_y = (th_max_y > height) ? height : th_max_y;
-		for (auto y = start_y; y < max_y; ++y) {
-		  for (auto x = start_x; x < max_x; ++x) {
-			render_pixel<width, height, samples, depth>(x, y, cam, hittable_ptr,
-														nb_hittable, fb_ptr);
-		  }
-		}
-	  });
-	});
+    auto fb_acc = frame_buf.get_access<sycl::access::mode::discard_write>(cgh);
+    auto hittables_acc =
+        hittables_buf.get_access<sycl::access::mode::read>(cgh);
+    hittable_t const* hittable_ptr = hittables_acc.get_pointer();
+    color* fb_ptr = fb_acc.get_pointer();
+    const auto global_range = sycl::range<1>(required_pg);
+    const auto local_range = sycl::range<1>(wg_size);
+    cgh.parallel_for(
+        sycl::nd_range<1>(global_range, local_range), [=](sycl::nd_item<1> it) {
+          auto gid = it.get_global_id(0);
+          const auto grid_x = gid % nb_pg_line;
+          const auto grid_y = gid / nb_pg_line;
+          const auto start_x = grid_x * pg_width;
+          const auto start_y = grid_y * pg_height;
+          const auto max_x = start_x + pg_width;
+          const auto th_max_y = start_y + pg_height;
+          const auto max_y = (th_max_y > height) ? height : th_max_y;
+          for (auto y = start_y; y < max_y; ++y) {
+            for (auto x = start_x; x < max_x; ++x) {
```

Comment:
> Why these loop nests in the case of a parallel_for ?

On my machine, using one thread per pixel lead to an under-utilization of the cpu (each core was having an activity between 70-80 %) because of scheduling issues.

On the gpu if we have more cores than pixel it should "degenerate" in using one core per work item (but I think the case is not handled well : the find immediate divider should return 1 if it reaches end of the function instead of zero).

Reply:
I'm waiting for your comment on this strategy before commenting.

On FPGA I think the problem is more complicated because you will want to "tune" the replication of the loop heart to find a good trade-off between area (to be able to pack other kernel on the same part without requiring reconfiguration) and execution time.

Reply:
There should not be one thread per pixel.
By looking at the code again, the problem is perhaps there is an explicit `nd_range` distribution of the parallelism. On CPU this is painful because you need to fight against possible barriers. This is why there is a macro to swear there is no barriers in use...
What about just using a simple `range` `parallel_for` and just trusting the runtime for distributing the work?
For triSYCL you could try the TBB runtime too.

Reply:
> By looking at the code again, the problem is perhaps there is an explicit nd_range distribution of the parallelism.

Indeed, there was an explicit nd range creating 64 work item per group organised in a 8x8 grid. I wasn't aware of 

> What about just using a simple range parallel_for and just trusting the runtime for distributing the work?

Ok that seems even better indeed. Done in 32baf38

Reply:
However, given this alternative API I don't see why there is a need of a "one_single_task" version for FPGA : the semantics of parallel_for should be sufficient, the backend compiler would then be responsible of choosing wether or not replicate many time the data flow or use loops.

Reply:
Sure. But the `parallel_for` might be less efficient on FPGA.